fix: Transfer ownership with S3 as primary

artonge opened the pull request:
## Issue

When using S3 as primary storage, transferring ownership with the `--move` option fail with the following error:

`SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry '8-45b963397aa40d4a0063e0d85e4fe7a1' for key 'fs_storage_path_hash'`

The `--move` option moves the entire home folder from one account to another.
The error means that the move failed because the destination folder already exist in `oc_filecache`.

Might have been introduced by: https://github.com/nextcloud/server/pull/26149

## Investigation

1. With S3 as primary storage, folders only exists as entries in `oc_filecache`.
2. With S3 as primary storage, `moveFromStorage(...)` only moves the cache entry, as nothing needs to be moved on disk. This cache move does not delete potentially pre-existing destination folder.
3. With Local storage, `moveFromStorage(...)` calls `rename(...)` which delete pre-existing folder.

#### S3's `moveFromStorage`

https://github.com/nextcloud/server/blob/373107b6e4ebfd3487c69a122ff3230057a4ae51/lib/private/Files/ObjectStore/ObjectStoreStorage.php#L593-L599

#### Local's `rename`, called by `moveFromStorage`

https://github.com/nextcloud/server/blob/373107b6e4ebfd3487c69a122ff3230057a4ae51/lib/private/Files/Storage/Local.php#L346-L352

## Solution

Delete pre-existing folder in `moveFromStorage(...)`

**Does it make sense to have that here? It feels off.**

## Todo

- [x] Target master
- [x] Add tests to `tests/lib/Files/Storage/StoragesTest.php`

artonge commented:
/backport to stable31

artonge commented:
/backport to stable30

artonge commented:
/backport to stable29

icewind1991 reviewed (changes requested):
The existing objects will have to be deleted to, not just the cache entries

artonge commented:
> The existing objects will have to be deleted to, not just the cache entries

Indeed, that's a concern. Not sure how to best proceed, because in case of a folder, we'll need to get all its children, no?
It leads me to wonder how are the children of a folder removed from the cache.

susnux commented on lib/private/Files/ObjectStore/ObjectStoreStorage.php:
```diff
@@ -592,7 +592,13 @@ public function copyFromStorage(
 		$sourceCache = $sourceStorage->getCache();
-		if ($sourceStorage->instanceOfStorage(ObjectStoreStorage::class) && $sourceStorage->getObjectStore()->getStorageId() === $this->getObjectStore()->getStorageId()) {
+		if (
+			$sourceStorage->instanceOfStorage(ObjectStoreStorage::class) &&
+			$sourceStorage->getObjectStore()->getStorageId() === $this->getObjectStore()->getStorageId()
```
```suggestion
			$sourceStorage->instanceOfStorage(ObjectStoreStorage::class)
			&& $sourceStorage->getObjectStore()->getStorageId() === $this->getObjectStore()->getStorageId()
```

IIRC thats our codestyle?

artonge replied:
But it is not aligned :(

artonge committed:
fix: Transfer ownership with S3 as primary

When using S3 as primary storage, transferring ownership with the `--move` option fail with the following error:

`SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry '8-45b963397aa40d4a0063e0d85e4fe7a1' for key 'fs_storage_path_hash'`

The `--move` option moves the entire home folder from one account to another.
The error means that the move failed because the destination folder already exist in `oc_filecache`.

- With S3 as primary storage, folders only exists as entries in `oc_filecache`.
- With S3 as primary storage, `moveFromStorage(...)` only moves the cache entry, as nothing needs to be moved on disk. This cache move does not delete potentially pre-existing destination folder.
- With Local storage, `moveFromStorage(...)` calls `rename(...)` which delete pre-existing folder.

- `transfer(...)`: https://github.com/nextcloud/server/blob/687a4d9ac7fcdbd935f81a0def567a1092306f7a/apps/files/lib/Service/OwnershipTransferService.php#L112
- `oneTimeUserSetup(...)`: https://github.com/nextcloud/server/blob/687a4d9ac7fcdbd935f81a0def567a1092306f7a/lib/private/Files/SetupManager.php#L261-L262
- `mkdir(...)`: https://github.com/nextcloud/server/blob/687a4d9ac7fcdbd935f81a0def567a1092306f7a/lib/private/Files/ObjectStore/ObjectStoreStorage.php#L91-L135
- `moveFromStorage(...)`: https://github.com/nextcloud/server/blob/687a4d9ac7fcdbd935f81a0def567a1092306f7a/lib/private/Files/ObjectStore/ObjectStoreStorage.php#L635-L636

Delete pre-existing folder in `moveFromStorage(...)`

Signed-off-by: Louis Chemineau <louis@chmn.me>

artonge commented:
> The existing objects will have to be deleted to, not just the cache entries

I added `$this->unlink($targetInternalPath);` and some extra assertion in the test.
